docs/MA/viergewinnt6.cpp

```diff
@@ -0,0 +1,781 @@
+public:
+    MinimaxNode() : value(0), value_not_none(false), children(), board(), player_am_zug(0), token(0), depth(0), expanded(false) {}
+
+    int value;
+    bool value_not_none;
+    std::vector<MinimaxNode> children;
+    std::vector<std::vector<int>> board;
+    int player_am_zug;
+    int token;
+    int depth;
+    bool expanded;
+
+    std::vector<MinimaxNode> expand_node() {
+        std::vector<MinimaxNode> new_children;
+        std::list<std::vector<std::vector<int>>> list_of_positions = generate_children(this->board, this->player_am_zug);
+        for (const std::vector<std::vector<int>>& board_position : list_of_positions) {
+            MinimaxNode child;
+            child.board = board_position;
+            child.player_am_zug = -this->player_am_zug;
+            child.token = this->token;
+            child.depth = this->depth + 1;
+            child.value_not_none = false;
+            child.value=0;
+            child.children;
+            new_children.push_back(child);
+
+        }
+        return new_children;
+    }
+
+    int minimax(int alpha, int beta, bool max_player, const int max_depth) {
+        //
+        minimax_counter+=1;
+        //
+        if (this->depth==max_depth || gewonnen(this->board,1) || gewonnen(this->board,-1) || game_over(this->board)) {
+            this->value = evaluate_position(this->board, this->token);
+            this->value_not_none=true;
+            return this->value;
+        }
+        //
+        if (!this->expanded) {this->children=this->expand_node(); this->expanded = true;}
+        //
+        if (max_player) {
+            int max_value=-1000000;
+            for (MinimaxNode& child : this->children) {
+                int eval=child.minimax(alpha,beta,false,max_depth);
+                if (eval>max_value) {max_value=eval;}
+                //pruning
+                if (eval>alpha) {alpha=eval;}
+                if (beta<=alpha) {break;}
+            }
+            this->value=max_value;
+            return max_value;
+        }
+        else if (!max_player) {
+            int min_value=1000000;
+            for (MinimaxNode& child : this->children) {
+                int eval=child.minimax(alpha,beta,true,max_depth);
+                if (eval<min_value) {min_value=eval;}
+                //pruning
+                if (eval<beta) {beta=eval;}
+                if (beta<=alpha) {break;}
+            }
+            this->value=min_value;
+            return min_value;
+        }
+
+    }
+
+    void sort(bool max_player) {
+        if (this->expanded) {
+            //
+            std::vector<MinimaxNode> not_none_children;
+            std::list<MinimaxNode> none_children;
+            std::vector<MinimaxNode> sorted_children;
+            //
+            for (const MinimaxNode& child : this->children) {
+                if (!child.value_not_none) {none_children.push_back(child);}
+                else if (child.value_not_none) {not_none_children.push_back(child);}
```

Comment:
```suggestion
                if (!child.value_not_none) {none_children.push_back(child);}
                else {not_none_children.push_back(child);}
```
So sparen Sie sich einen Vergleich